expense_tracker/lib/expenses_screen/expenses_list/new_expense.dart

```diff
@@ -0,0 +1,244 @@
+            icon: const Icon(Icons.error),
+            actions: [
+              TextButton(
+                onPressed: () {
+                  Navigator.pop(ctx);
+                },
+                child: const Text('Okay'),
+              ),
+            ],
+          );
+        },
+      );
+    } else {
+      final Expense newExpense = Expense(
+          title: _titleController.text.trim(),
+          category: selectedCategory,
+          amount: amount,
+          date: selectedDate);
+
+      Expense? expenseToUpdate = widget.expenseToUpdate;
+      if (expenseToUpdate != null) {
+        widget.onUpdateExpense!(
+            oldExpense: expenseToUpdate, updatedExpense: newExpense);
+      } else {
+        widget.onAddExpense!(newExpense);
+      }
+      Navigator.pop(context);
+    }
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    // Fill fields with details of existinf expense if provided
+    Expense? expenseToUpdate = widget.expenseToUpdate;
+    if (expenseToUpdate != null) {
+      _amountController.text = expenseToUpdate.amount.toString();
+      _titleController.text = expenseToUpdate.title;
+      selectedCategory = expenseToUpdate.category;
+      selectedDate = expenseToUpdate.date;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final keyboardSpace = MediaQuery.of(context).viewInsets.bottom;
+
+    // Reusable  widgets extracted for readability and no duplication
+    var cancelButton = TextButton(
+        onPressed: () {
+          Navigator.pop(context);
+        },
+        child: const Text('Cancel'));
+
+    var saveExpenseButton = ElevatedButton(
+        onPressed: _submitExpenceData, child: const Text('Save Expense'));
+
+    var titleTextField = TextField(
+      maxLength: 50,
+      keyboardType: TextInputType.text,
+      controller: _titleController,
+      decoration: const InputDecoration(label: Text('Title')),
+    );
+
+    var amountTextField = TextField(
+        keyboardType: TextInputType.number,
+        controller: _amountController,
+        decoration:
+            const InputDecoration(label: Text('Amount'), prefixText: '\$'));
+
+    var dropdownCategorySelectionButton = DropdownButton(
+      value: selectedCategory,
+      items: Category.values
+          .map((item) => DropdownMenuItem(
+              value: item,
+              child: Text(
+                item.name.toUpperCase(),
+              )))
+          .toList(),
+      onChanged: (value) {
+        if (value == null) {
+          return;
+        }
+        setState(
+          () {
+            selectedCategory = value;
+          },
+        );
+      },
+    );
+
+    var datePickerRow = Row(
+      mainAxisAlignment: MainAxisAlignment.end,
+      children: [
+        Text(formatter.format(selectedDate)),
+        IconButton(
+          onPressed: _presentDatePicker,
+          icon: const Icon(
+            Icons.calendar_month,
+          ),
+        ),
+      ],
+    );
```

Comment:
Nice to have this seperated like this, makes the code much more readable when you re-use the components.